File: Python/services.py
import os
import requests
from dotenv import load_dotenv

load_dotenv()

API_URL = os.getenv("API_URL", "")
API_KEY = os.getenv("API_KEY", "")

if not API_URL:
    print("⚠️ Warning: API_URL is not set in .env")
if not API_KEY:
    print("⚠️ Warning: API_KEY is not set in .env")

def generate_scoped_public_key(engines, expires_in_seconds):
    try:
        payload = {}
        if engines is not None:
            payload["engines"] = engines
        if expires_in_seconds is not None:
            payload["expiresInSeconds"] = expires_in_seconds

        headers = {
            "Content-Type": "application/json",
            "X-Api-Key": API_KEY
        }

        response = requests.post(
            f"{API_URL}/plan/auth/generate-scoped-public-key",
            json=payload,
            headers=headers,
            timeout=30
        )

        response.raise_for_status()

        print("✅ Generate Scoped Public Key API Request Successful!")

        return response.json()

    except requests.exceptions.HTTPError as e:
        print("❌ Generate Scoped Public Key API Request Failed!", e.response)

        status_code = e.response.status_code

        print(f"Error Status Code: {status_code}")
        print(f"Error Details: {e.response.text}")

        if status_code == 401:
            print("💡 Hint: API Key is missing or invalid.")
        elif status_code == 403:
            print("💡 Hint: Your IP address is not included in the IP allowlist.")

        return None

    except requests.exceptions.RequestException as e:
        print(
            "No response received from server. "
            "Please check your URL or ensure the backend server is running."
        )
        print(e)
        return None

File: PHP/services.php
<?php

function generateScopedPublicKey(?array $engines = null, ?int $expiresInSeconds = null)
{
    $apiUrl = getenv('API_URL');
    $apiKey = getenv('API_KEY');

    if (!$apiUrl || !$apiKey) {
        error_log('API_URL or API_KEY is not configured.');
        return null;
    }

    $data = [];
    if ($engines !== null) {
        $data['engines'] = $engines;
    }
        if ($expiresInSeconds !== null) {
        $data['expiresInSeconds'] = $expiresInSeconds;
    }

    $payload = json_encode($data, JSON_FORCE_OBJECT);

    $ch = curl_init();

    curl_setopt_array($ch, [
        CURLOPT_URL => rtrim($apiUrl, '/') . '/plan/auth/generate-scoped-public-key',
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_POST => true,
        CURLOPT_POSTFIELDS => $payload,
        CURLOPT_HTTPHEADER => [
            'Content-Type: application/json',
            'X-Api-Key: ' . $apiKey
        ],
        CURLOPT_TIMEOUT => 30
    ]);

    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);

    if (curl_errno($ch)) {
        error_log('Request failed: ' . curl_error($ch));
        curl_close($ch);
        return null;
    }

    curl_close($ch);

    if ($httpCode < 200 || $httpCode >= 300) {
        error_log("Generate Scoped Public Key API failed. HTTP {$httpCode}");
        error_log($response);

        if ($httpCode === 401) {
            error_log('Hint: API Key is missing or invalid.');
        } elseif ($httpCode === 403) {
            error_log('Hint: Your IP address is not included in the IP allowlist.');
        }

        return null;
    }

    return json_decode($response, true);
}
